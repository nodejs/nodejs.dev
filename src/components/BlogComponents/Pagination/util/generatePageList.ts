import { PageModel, PageType } from '../../../../types/pagination';

export function generatePageList(
  pageCount: number,
  currentPage: number,
  showPages: boolean
): PageModel[] {
  const pages: PageModel[] = [];
  if (showPages) {
    for (let i = 1; i <= pageCount; i += 1) {
      const selected = i === currentPage;
      pages.push({
        type: PageType.NUM,
        num: i,
        selected,
      });
    }
  }

  return pages;
}
