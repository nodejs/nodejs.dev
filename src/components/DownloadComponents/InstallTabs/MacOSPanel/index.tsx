import React from 'react';
import { FormattedMessage } from 'react-intl';
import { LocalizedLink as Link } from 'gatsby-theme-i18n';
import useNvmVersion from '../../../../hooks/useNvmVersion';
import { ShellBox } from '../../../CommonComponents';
import styles from '../index.module.scss';

interface Props {
  nvmVersion: string;
}

export const PureMacOSPanel = ({ nvmVersion }: Props): JSX.Element => {
  const nvmInstallScriptUrl = `https://raw.githubusercontent.com/nvm-sh/nvm/${nvmVersion}/install.sh`;

  return (
    <div>
      <ShellBox textToCopy={`curl -o- ${nvmInstallScriptUrl} | bash`}>
        <span className={styles.installTextNoSelect}>$</span>
        <span className={styles.installTextCommand}>curl -o- </span>
        {nvmInstallScriptUrl}
        <span className={styles.installTextCommand}> | bash </span>
      </ShellBox>
      <ShellBox textToCopy="nvm install --lts">
        <span className={styles.installTextNoSelect}>$</span>
        <span className={styles.installTextCommand}>nvm </span>install --lts
      </ShellBox>
      <br />
      <br />
      <Link
        className={styles.installDocsButton}
        to="/download/package-manager/#nvm"
      >
        <FormattedMessage id="components.installTabs.readDocs" />
      </Link>
    </div>
  );
};

const MacOSPanel = (): JSX.Element => {
  const nvmVersion = useNvmVersion();

  return <PureMacOSPanel nvmVersion={nvmVersion} />;
};

export default MacOSPanel;
