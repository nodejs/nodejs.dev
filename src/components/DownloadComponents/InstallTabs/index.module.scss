.install {
  .react-tabs {
    -webkit-tap-highlight-color: transparent;

    &__tab-list {
      display: flex;
      margin: 0;
      padding: 0;
      text-align: center;
    }

    &__tab {
      background: var(--black11);
      border: 1px solid transparent;
      border-bottom: none;
      bottom: -1px;
      color: var(--black4);
      cursor: pointer;
      display: inline-block;
      list-style: none;
      padding: 10px 12px;
      position: relative;
      text-align: center;
      width: 100%;

      &--selected {
        background: var(--black10);
        border-radius: 5px 5px 0 0;
        color: #fff;
      }

      &:focus {
        border-bottom: none;
        border-color: var(--pink5);
        box-shadow: 0 0 5px var(--pink5);
        outline: none;

        &:after {
          background: var(--black10);
          bottom: -5px;
          content: '';
          height: 5px;
          left: 0;
          position: absolute;
          right: 0;
        }
      }
    }

    &__tab-panel {
      display: none;
      padding: 20px;

      &--selected {
        display: block;
      }
    }
  }

  &__header {
    font-weight: var(--font-weight-semibold);
    height: 30px;
    margin: 0 5px;

    &-circles {
      float: left;
      height: 100%;
    }

    &-grey-circle {
      background-color: var(--black8);
      border-radius: 50%;
      display: inline-block;
      height: 10px;
      margin: 10px 5px;
      width: 10px;
    }

    &-text {
      line-height: 30px;
      text-align: center;
      width: calc(100% - 60px);
    }
  }

  &__text {
    color: var(--pink5);

    &__line {
      margin: var(--space-12) 0;
    }

    &__no-select {
      margin-right: var(--space-12);
      user-select: none;
    }

    &__command {
      color: var(--black0);
    }
  }

  &__docs-button,
  &__docs-button:hover {
    border: 1px solid var(--black3);
    border-radius: var(--space-04);
    bottom: 2rem;
    box-sizing: border-box;
    color: var(--black3);
    font-weight: var(--font-weight-semibold);
    height: 4rem;
    padding: var(--space-08) var(--space-16);
    position: absolute;
    right: 2rem;
    text-decoration: none;
  }
}
