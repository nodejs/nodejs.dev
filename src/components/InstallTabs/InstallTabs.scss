.react-tabs {
  -webkit-tap-highlight-color: transparent;

  &__tab-list {
    margin: 0;
    padding: 0;
    text-align: center;
    display: flex;
  }

  &__tab {
    display: inline-block;
    border: 1px solid transparent;
    border-bottom: none;
    bottom: -1px;
    color: var(--black4);
    position: relative;
    list-style: none;
    padding: 10px 12px;
    background: var(--black8);
    cursor: pointer;
    width: 100%;
    text-align: center;

    &--selected {
      background: var(--black9);
      color: #fff;
      border-radius: 5px 5px 0 0;
    }

    &:focus {
      box-shadow: 0 0 5px var(--pink5);
      border-color: var(--pink5);
      border-bottom: none;
      outline: none;

      &:after {
        content: '';
        position: absolute;
        height: 5px;
        left: 0;
        right: 0;
        bottom: -5px;
        background: var(--black9);
      }
    }
  }

  &__tab-panel {
    display: none;
    padding: 20px;

    &--selected {
      display: block;
    }
  }

  .install {
    &__header {
      height: 30px;
      margin: 0 5px;
      font-weight: var(--font-weight-semibold);
      &-circles {
        height: 100%;
        float: left;
      }
      &-grey-circle {
        display: inline-block;
        background-color: var(--black8);
        width: 10px;
        height: 10px;
        border-radius: 50%;
        margin: 10px 5px;
      }
      &-text {
        width: calc(100% - 60px);
        text-align: center;
        line-height: 30px;
      }
    }
    &__text {
      color: var(--pink5);
      &__line {
        margin: var(--space-12) 0;
      }
      &__no-select {
        margin-right: var(--space-12);
        -webkit-touch-callout: none; /* iOS Safari */
        -webkit-user-select: none; /* Safari */
        -khtml-user-select: none; /* Konqueror HTML */
        -moz-user-select: none; /* Old versions of Firefox */
        -ms-user-select: none; /* Internet Explorer/Edge */
        user-select: none;
      }
    }
    &__docs-button,
    &__docs-button:hover {
      border-radius: var(--space-04);
      border: 1px solid var(--color-text-secondary);
      padding: var(--space-08) var(--space-16);
      position: absolute;
      bottom: 2rem;
      right: 2rem;
      text-decoration: none;
      box-sizing: border-box;
      height: 4rem;
      color: var(--color-text-secondary);
      font-weight: var(--font-weight-semibold);
    }
  }
  .shell-box {
    code {
      color: var(--pink5);
      width: calc(100% - 20px);
      line-height: 30px;
      > span.install__text__command {
        color: #fff;
      }
    }
    background-color: var(--black9);
    padding-bottom: 15px;
  }
}
