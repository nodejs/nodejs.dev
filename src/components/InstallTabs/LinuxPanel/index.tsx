import React from 'react';
import ShellBox from '../../ShellBox';
import '../InstallTabs.scss';

const LinuxPanel = (): JSX.Element => {
  return (
    <div>
      <ShellBox
        textToCopy="apk add -U curl
bash ca-certificates openssl ncurses coreutils python2 make gcc
g++ libgcc linux-headers grep util-linux binutils findutils"
      >
        <span className="install__text__no-select">$</span>
        <span className="install__text__command">apk add -U</span> curl bash
        ca-certificates openssl ncurses coreutils python2 make gcc g++ libgcc
        linux-headers grep util-linux binutils findutils
      </ShellBox>
      <ShellBox textToCopy="curl -o- | bash https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh">
        <span className="install__text__no-select">$</span>
        <span className="install__text__command"> curl -o- | bash </span>
        https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh
      </ShellBox>
      <ShellBox textToCopy="nvm install --lts">
        <span className="install__text__no-select">$</span>
        <span className="install__text__command"> nvm</span> install --lts
      </ShellBox>
      <br />
      <br />
      {/* TODO when the new docs page is ready link to that page.  */}
      <a
        className="install__docs-button"
        href="https://nodejs.org/en/download/package-manager/#nvm"
        target="_blank"
        rel="noopener noreferrer"
      >
        Read documentation
      </a>
    </div>
  );
};

export default LinuxPanel;
