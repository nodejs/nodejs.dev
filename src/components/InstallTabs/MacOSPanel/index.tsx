import React from 'react';
import ShellBox from '../../ShellBox';
import '../InstallTabs.scss';

const MacOSPanel = (): JSX.Element => {
  return (
    <div>
      <ShellBox textToCopy="curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.37.2/install.sh | bash">
        <span className="install__text__no-select">$</span>
        <span className="install__text__command">curl -o- </span>
        https://raw.githubusercontent.com/nvm-sh/nvm/v0.37.2/install.sh
        <span className="install__text__command"> | bash </span>
      </ShellBox>
      <ShellBox textToCopy="nvm install --lts">
        <span className="install__text__no-select">$</span>
        <span className="install__text__command">nvm </span>install --lts
      </ShellBox>
      <br />
      <br />
      <a
        className="install__docs-button"
        href="https://nodejs.org/en/download/package-manager/#nvm"
        target="_blank"
        rel="noopener noreferrer"
      >
        Read documentation
      </a>
    </div>
  );
};

export default MacOSPanel;
