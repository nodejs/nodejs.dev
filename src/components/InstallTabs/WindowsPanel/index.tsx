import React from 'react';
import ShellBox from '../../ShellBox';
import '../InstallTabs.scss';

const WindowsPanel = (): JSX.Element => {
  return (
    <div>
      <ShellBox textToCopy="choco install nodejs-lts">
        <span className="install__text__no-select">$</span>
        <span className="install__text__command">choco </span>install nodejs-lts
      </ShellBox>
      <br />
      <br />
      <a
        className="install__docs-button"
        href="https://nodejs.org/en/download/package-manager/#windows"
        target="_blank"
        rel="noopener noreferrer"
      >
        Read documentation
      </a>
    </div>
  );
};

export default WindowsPanel;
