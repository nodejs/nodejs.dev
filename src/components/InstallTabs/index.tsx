import React from 'react';
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import { detectOS, UserOS } from '../../util/detectOS';
import './InstallTabs.scss';

import WindowsPanel from './WindowsPanel';
import MacOSPanel from './MacOSPanel';
import LinuxPanel from './LinuxPanel';

const InstallTabs = (): JSX.Element => {
  const userOS = detectOS();

  const os = {
    win: 'Windows (Chocolatey)',
    mac: 'macOS (nvm)',
    linux: 'Linux (nvm)',
  };

  const installTabSystems = {
    WIN: [os.win, os.mac, os.linux],
    MAC: [os.mac, os.win, os.linux],
    LINUX: [os.linux, os.mac, os.win],
    UNIX: [os.linux, os.mac, os.win],
    UNKNOWN: [os.win, os.mac, os.linux],
  };

  function panelSwitch(): JSX.Element {
    switch (userOS) {
      case UserOS.MAC:
        return (
          <>
            <TabPanel>
              <MacOSPanel />
            </TabPanel>
            <TabPanel>
              <WindowsPanel />
            </TabPanel>
            <TabPanel>
              <LinuxPanel />
            </TabPanel>
          </>
        );
      case UserOS.LINUX:
      case UserOS.UNIX:
        return (
          <>
            <TabPanel>
              <LinuxPanel />
            </TabPanel>
            <TabPanel>
              <MacOSPanel />
            </TabPanel>
            <TabPanel>
              <WindowsPanel />
            </TabPanel>
          </>
        );
      default:
        return (
          <>
            <TabPanel>
              <WindowsPanel />
            </TabPanel>
            <TabPanel>
              <MacOSPanel />
            </TabPanel>
            <TabPanel>
              <LinuxPanel />
            </TabPanel>
          </>
        );
    }
  }

  return installTabSystems[userOS] !== undefined ? (
    <div className="install">
      <Tabs>
        <div className="install__header">
          <div className="install__header-circles">
            <div className="install__header-grey-circle" />
            <div className="install__header-grey-circle" />
            <div className="install__header-grey-circle" />
          </div>
          <div className="install__header-text">
            {userOS === 'MAC' ? 'zsh' : 'bash'}
          </div>
        </div>
        <TabList>
          {installTabSystems[userOS].map((system: string) => (
            <Tab key={system.toString()}>{system}</Tab>
          ))}
        </TabList>
        {panelSwitch()}
      </Tabs>
    </div>
  ) : (
    <></>
  );
};

export default InstallTabs;
