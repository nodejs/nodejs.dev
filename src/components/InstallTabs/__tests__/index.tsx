import React from 'react';
import { render } from '@testing-library/react';
import InstallTabs from '../index';
import * as detectOSModule from '../../../util/detectOS';
import { UserOS } from '../../../util/detectOS';

jest.mock('../../../util/detectOS');

describe('Tests for InstallTabs component', () => {
  it('renders correctly', () => {
    jest.spyOn(detectOSModule, 'detectOS').mockReturnValue(UserOS.UNKNOWN);
    const { container } = render(<InstallTabs />);
    expect(container).toMatchSnapshot();
  });
  it('renders correctly for macOS', () => {
    jest.spyOn(detectOSModule, 'detectOS').mockReturnValue(UserOS.MAC);
    const { container } = render(<InstallTabs />);
    expect(container).toMatchSnapshot();
  });
  it('renders correctly for Linux', () => {
    jest.spyOn(detectOSModule, 'detectOS').mockReturnValue(UserOS.LINUX);
    const { container } = render(<InstallTabs />);
    expect(container).toMatchSnapshot();
  });
  it('renders correctly for Unix', () => {
    jest.spyOn(detectOSModule, 'detectOS').mockReturnValue(UserOS.UNIX);
    const { container } = render(<InstallTabs />);
    expect(container).toMatchSnapshot();
  });
  it('renders correctly for other', () => {
    jest.spyOn(detectOSModule, 'detectOS').mockReturnValue(UserOS.WIN);
    const { container } = render(<InstallTabs />);
    expect(container).toMatchSnapshot();
  });
  it('renders correctly for mobile', () => {
    jest.spyOn(detectOSModule, 'detectOS').mockReturnValue(UserOS.MOBILE);
    const { container } = render(<InstallTabs />);
    expect(container).toMatchSnapshot();
  });
});
