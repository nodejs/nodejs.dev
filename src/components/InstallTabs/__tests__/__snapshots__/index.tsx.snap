// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for InstallTabs component renders correctly 1`] = `
<div>
  <div
    class="install"
  >
    <div
      class="react-tabs"
      data-tabs="true"
    >
      <div
        class="install__header"
      >
        <div
          class="install__header-circles"
        >
          <div
            class="install__header-grey-circle"
          />
          <div
            class="install__header-grey-circle"
          />
          <div
            class="install__header-grey-circle"
          />
        </div>
        <div
          class="install__header-text"
        >
          bash
        </div>
      </div>
      <ul
        class="react-tabs__tab-list"
        role="tablist"
      >
        <li
          aria-controls="react-tabs-1"
          aria-disabled="false"
          aria-selected="true"
          class="react-tabs__tab react-tabs__tab--selected"
          id="react-tabs-0"
          role="tab"
          tabindex="0"
        >
          Windows (Chocolatey)
        </li>
        <li
          aria-controls="react-tabs-3"
          aria-disabled="false"
          aria-selected="false"
          class="react-tabs__tab"
          id="react-tabs-2"
          role="tab"
        >
          macOS (nvm)
        </li>
        <li
          aria-controls="react-tabs-5"
          aria-disabled="false"
          aria-selected="false"
          class="react-tabs__tab"
          id="react-tabs-4"
          role="tab"
        >
          Linux (nvm)
        </li>
      </ul>
      <div
        aria-labelledby="react-tabs-0"
        class="react-tabs__tab-panel react-tabs__tab-panel--selected"
        id="react-tabs-1"
        role="tabpanel"
      >
        <div>
          <pre
            class="shell-box"
          >
            <div
              class="shell-box-top"
            >
              <span>
                SHELL
              </span>
              <button
                type="button"
              >
                copy
              </button>
            </div>
            <code>
              <span
                class="install__text__no-select"
              >
                $
              </span>
              <span
                class="install__text__command"
              >
                choco 
              </span>
              install nodejs-lts
            </code>
          </pre>
          <br />
          <br />
          <a
            class="install__docs-button"
            href="https://nodejs.org/en/download/package-manager/#windows"
            rel="noopener noreferrer"
            target="_blank"
          >
            Read documentation
          </a>
        </div>
      </div>
      <div
        aria-labelledby="react-tabs-2"
        class="react-tabs__tab-panel"
        id="react-tabs-3"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-4"
        class="react-tabs__tab-panel"
        id="react-tabs-5"
        role="tabpanel"
      />
    </div>
  </div>
</div>
`;
